File: regional/regional/2015/I_O/Gears/Gears.py
import sys
from fractions import gcd
N = int(raw_input())
gears = []
for i in range(N):
   gears.append([int(j) for j in raw_input().split()]) # x y r
dir = [0 for j in range(N)]
dir[0] = 1
q = []
q.append(0)
qget = 0
locked = False
def dist2(x, y):
   return x * x + y * y
while qget < len(q):
   i = q[qget]
   qget += 1
   for j in range(N):
      if i != j:
         d = dist2(gears[i][0]-gears[j][0], gears[i][1]-gears[j][1])
         rr = dist2(gears[i][2]+gears[j][2], 0)
         if d < rr:
            print "Dist", d
            print "rr", rr
            print gears[i][0], gears[i][1], gears[i][2] ;
            print gears[j][0], gears[j][1], gears[j][2] ;
            sys.exit("Gears not identical but too close")
         elif d == rr:
            if dir[j] == 0:
               dir[j] = - dir[i] ;
               q.append(j)
            elif dir[j] == dir[i]:
               locked = True
if locked:
   print "The input gear cannot move."
elif dir[N-1] == 0:
   print "The input gear is not connected to the output gear."
else:
   g = gcd(gears[0][2], gears[N-1][2])
   print str(dir[N-1]*gears[0][2]/g)+":"+str(gears[N-1][2]/g)

File: regional/PacificNW/2015/I_O/Gears/Gears.java
import java.util.*;
public class Gears {
	static int[] x, y, r;
	public static void main(String[] args) throws Exception {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		x = new int[n];
		y = new int[n];
		r = new int[n];
		for(int i = 0; i < n; i++) {
			x[i] = sc.nextInt();
			y[i] = sc.nextInt();
			r[i] = sc.nextInt();
		}
		LinkedList<Integer> q = new LinkedList<Integer>();
		int[] dp = new int[n];
		dp[0] = 1;
		boolean locked = false;
		q.add(0);
		while(!q.isEmpty()) {
			int curr = q.removeFirst();
			for(int i = 0; i < n; i++) {
				if(i == curr) continue;
				if(connected(curr, i)) {
					if(dp[i] == dp[curr]) {
						locked = true;
					}
					else if(dp[i] == 0) {
						dp[i] = -dp[curr];
						q.add(i);
					}
				}
			}
		}
		if(locked) {
			System.out.println("The input gear cannot move.");
		}
		else if(dp[n-1] == 0) {
			System.out.println("The input gear is not connected to the output gear.");
		}
		else {
			int gcd = gcd(r[n-1], r[0]);
			System.out.println(dp[n-1]*r[0]/gcd + ":" + r[n-1]/gcd);
		}
	}
	public static boolean connected(int i, int j) {
		int xx = x[i] - x[j];
		int yy = y[i] - y[j];
		int rr = xx*xx+yy*yy;
		int other = r[i] + r[j];
		other *= other;
		return rr==other;
	}
	public static int gcd(int a, int b) {
		return a%b==0 ? b : gcd(b, a%b);
	}
}
